### PR TITLE
Fix #1681 mount: change behavior of `restic mount` when path doesn't exist

### LINKED ISSUE #1681: Restic fails to create the mount point directory if the parent directory does not exist
## Output of `restic version`
`restic 0.8.3
compiled with go1.10 on darwin/amd64`

## How did you run restic exactly?
`sudo restic -r /Volumes/ex mount /mnt/bak `

Since this is a Mac `/mnt` does not exist yet.

## What backend/server/service did you use to store the repository?
Local

## Expected behavior
If the mount point is not accessible to restic, one of:
- Restic would not try to create any directory and just fail if the mount point does not exist
- Restic would try and create the mount point and all parent directories

## Actual behavior
Restic wants to be nice and tries to create the mount point for you, if it notices that it's missing.
However restic does not check if the parent directory of the mount point exists. This leads to a situation where restic tries to create the mount point, but the operation fails due to the parent directory not existing.

## Steps to reproduce the behavior
```
# restic -r /tmp/test init
# restic -r /tmp/test mount $HOME/path/to/your/very/deep/mount/directory/thats/without/a/parent

enter password for repository: 
password is correct
Mountpoint /Users/nobody/path/to/your/very/deep/mount/directory/thats/without/a/parent doesn't exist, creating it
mkdir /Users/nobody/path/to/your/very/deep/mount/directory/thats/without/a/parent: no such file or directory
unable to umount (maybe already umounted?): unmount /Users/nobody/path/to/your/very/deep/mount/directory/thats/without/a/parent: no such file or directory
```

## Do you have any idea what may have caused this?
See Line: https://github.com/restic/restic/blob/be0a5b7f065a0b80f82c3ba1aa5abd9253b150ee/cmd/restic/cmd_mount.go#L103:18

## Do you have an idea how to solve the issue?
In this case I don't think restic has to be extra nice to the end user. The program should just fail if the mount point does not exist.

However if restic wants to be nice to the end user, we would need to add `MkDirAll` to the FS functions and use that to create the mount point.

### PR DESCRIPTION
**Pull request by kitone**
What does this PR change? What problem does it solve?
-----------------------------------------------------

The `restic mount` command will not try to create any directory and just fail if the mount point does not exist with explicit messages. The checks of the validity of the mount point (exist and permission) will be done sooner.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
#1681
~~#1724~~

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- ~~I have added tests for all changes in this PR~~
- ~~I have added documentation for the changes (in the manual)~~
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by rawtaz** [comment]
This changes the behavior of the `mount` command in that it no longer creates the mount point folder if it doesn't exist. If we decide to do this, which would break BC but is [suggested by @fd0 in #1681](https://github.com/restic/restic/issues/1681#issuecomment-375785383), we must be very clear about it in the changelog entry (which is needed regardless).

File: changelog/unreleased/issue-1681 [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Bugfix: mount: change behavior of `restic mount` when path doesn't exist
+
+We've change the behavior of `restic mount` command when the directory doesn't
+exist, prevously restic was trying to create the mount point path in that case,
+this is not the current behavior anymore, we now return an explicit error.
+In addition to this, we check that we have permission to write in the directory.
+
+
```
Comment (rawtaz) [inline changelog/unreleased/issue-1681:8]: Thanks for adding the changelog! Please update it to the following:

~~~
Bugfix: Make mount check permissions and not create missing mount point

When specifying a non-existent directory as mount point for the mount
command, restic used to create the directory automatically. This has now
changed such that restic instead gives an error when the directory for the
mount point doesn't exist.

Restic now also checks the permissions of the specified mount point to
make sure that they are sufficient for mounting the repository in it.
~~~

Also please remove the excessive empty line before the URLs at the end :)

Comment (kitone) [reply changelog/unreleased/issue-1681]: Thanks, done!

File: cmd/restic/cmd_mount.go [hunk 2]
```diff
@@ -160,6 +159,32 @@ func umount(mountpoint string) error {
...
+	if err != nil {
+		// check with IsPermission(err) ?
+		return errors.Fatalf("the mount path is not writable directory: %q", mountpoint)
+	}
+	// fixme: check for error ?
+	f.Close()
+	os.Remove(tmpFile)
+
+	return nil
+}
```
Comment (greatroar) [inline cmd/restic/cmd_mount.go:186]: I don't see why any of this is necessary. When the mountpoint doesn't exist, bazil.org/fuse reports that:

    mountpoint does not exist: /tmp/foo

When the permission are wrong, it reports that:

    2020/10/11 16:21:43 mount helper error: fusermount: user has no write access to mountpoint /tmp/foo

File: cmd/restic/cmd_mount.go [hunk 3]
```diff
@@ -104,11 +106,8 @@ func mount(opts MountOptions, gopts GlobalOptions, mountpoint string) error {
 	}
 
 	if _, err := resticfs.Stat(mountpoint); os.IsNotExist(errors.Cause(err)) {
-		Verbosef("Mountpoint %s doesn't exist, creating it\n", mountpoint)
-		err = resticfs.Mkdir(mountpoint, os.ModeDir|0700)
-		if err != nil {
-			return err
-		}
+		Verbosef("Mountpoint %s doesn't exist\n", mountpoint)
+		return err
```
Comment (greatroar) [inline cmd/restic/cmd_mount.go:110]: This check is redundant.

Comment (kitone) [reply cmd/restic/cmd_mount.go]: Thank, forget to remove it

File: cmd/restic/cmd_mount.go (hunk 2 above)
Comment (kitone) [reply cmd/restic/cmd_mount.go:180]: Checking about this earlier make it fail faster and remove unnecessary work and terminal output like opening repo, ...
For the end user, this is a more friendly error message that reading.

But I guess we can remove this commit and simply close the #1724, on my last test the message print without blocking.

Comment (MichaelEischer) [reply cmd/restic/cmd_mount.go:180]: Let's assume for now that #1724 was caused by something external to restic. So please remove this commit for now.

Comment (kitone) [reply cmd/restic/cmd_mount.go]: Hi, the commit was removed and the changelog updated.

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**